website/utils/datatype_utils.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Returns a (shallow) copy of an object with all properties that are
+ * `undefined` removed.
+ */
+export function removeUndefinedProperties<T>(obj: T): Partial<T> {
+  const result: Partial<T> = {};
+  for (const key in obj) {
+    if (obj[key] !== undefined) {
+      result[key] = obj[key];
+    }
+  }
+  return result;
+}
+
+function isArrayDeepEqual(a: unknown[], b: unknown[]): boolean {
+  if (!Array.isArray(a) || !Array.isArray(b)) {
+    return false;
+  }
+  if (a.length !== b.length) {
+    return false;
+  }
+
+  return a.every((val, i) => {
+    return isDeepEqual(val, b[i]);
+  });
+}
+
+function isObjectDeepEqual(a: Record<string, unknown>, b: Record<string, unknown>): boolean {
+  if (Object.keys(a).length !== Object.keys(b).length) {
+    return false;
+  }
+  for (const key in a) {
+    if (!isDeepEqual(a[key], b[key])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Recursively checks for deep equality between two elements, including objects or arrays.
+ * @returns true if `a` and `b` are deeply equal, false otherwise. Handles NaN values
+ * and nested arrays.
+ */
+export function isDeepEqual(a: unknown, b: unknown): boolean {
```

Comment:
see if the lodash deep equality check will work here

Reply:
It does! Test that uses it passes :)